Major Revision: combobox role (#14534)
mdn/content: files/en-us/web/accessibility/aria/roles/combobox_role/index.md

  - Reference
  - ARIA roles
  - composite widget role
  - combobox
---

The `combobox` role identifies an element as an `input` that controls another element, such as a `listbox` or `grid`, that can dynamically pop up to help the user set the value of that `input`.

## Description

A `combobox` is a composite widget that combines a named input field with a popup providing possible values for that input field. The purpose of a this widget is to improve user experience by helping the user select a value without having to type in the complete value and, optionally depending whether supported values are limited, preventing the user from entering invalid or otherwise unsupported values. 

The `combobox` role is set on input that controls another element, such as a listbox or grid, that can dynamically pop up to help the user set the value of the input. 

The `combobox` input field can either be a single-line text field that supports editing and typing, similar to a HTML {{HTMLElement('input')}} with a {{HTMLElement('datalist')}}, or an element that only displays the current value of the combobox. 

> A combobox only has one required attribute: [`aria-expanded`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded). Generally, it also has several other attributes including [`aria-haspopup`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup),  [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls), [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-activedescendant), and [`aria-autocomplete`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-autocomplete).

Reviewer: While the `aria-expanded` attribute is the only 'required' attribute, the others listed have implicit fallback values if authors fail to specify them.  these attributes are still "required" in a sense, but due to them having fallback values, or being contingent on the manner in which the combobox is implemented (e.g., activedescendant/aria-controls aren't "required" if an author specifically manages focus/uses a roving tabindex for an invoked listbox).

Here's a suggested update, but if you feel other wording would be better, then please do what you think is best.

Note I specifically said "A WAI-ARIA combobox" purely based on having to deal with auditors thinking that that an `<input list=foo><datalist id=foo>...</datalist>` also required an `aria-expanded` to be set.  which is not something that devs should be doing

```suggestion
A WAI-ARIA combobox only has one attribute that is required that authors specify: [`aria-expanded`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded). However, it also has several other attributes which will be necessary to specify, depending on the combobox's implementation. These include [`aria-haspopup`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup), [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls), [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-activedescendant), and [`aria-autocomplete`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-autocomplete).
```

> Typically, the initial state of a combobox is collapsed, with `aria-expanded="false"` set. In the collapsed state, only the combobox element and, optionally, a separate, optional, popup control button are visible. The [`aria-expanded`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded), with the value set to `false`, is required when collapsed, because it indicates to assistive technologies that the widget is expandable.

Reviewer: revised wording to remove the redundant use of 'optional'

```suggestion
Typically, the initial state of a combobox is collapsed, with `aria-expanded="false"` set. In the collapsed state, only the combobox element and, optionally a sibling button to invoke the popup, are visible. The [`aria-expanded`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded), with the value set to `false`, is required when collapsed, because it indicates to assistive technologies that the widget is expandable.
```

> The combobox is said to be expanded when both the combobox element showing its current value and its associated popup element are visible. When expanded, `aria-expanded="true"` must be set.

Reviewer: ```suggestion
The combobox is in the expanded state when both the combobox element showing its current value and its associated popup element are visible. When expanded, `aria-expanded="true"` must be set.
```

> The popup element associated with a `combobox` can be either a [`listbox`](/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role), [`tree`](/en-US/docs/Web/Accessibility/ARIA/Roles/tree_role), [`grid`](/en-US/docs/Web/Accessibility/ARIA/Roles/grid_role), or [`dialog`](/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) element, with `listbox` being the default type. 

Reviewer: Removal of mentioning the listbox is the default type, as the next paragraph repeats this.

```suggestion
The popup element associated with a `combobox` can be either a [`listbox`](/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role), [`tree`](/en-US/docs/Web/Accessibility/ARIA/Roles/tree_role), [`grid`](/en-US/docs/Web/Accessibility/ARIA/Roles/grid_role), or [`dialog`](/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) element. 
```

Comboboxes have an implicit [`aria-haspopup`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup) value of `listbox`, so including this attribute is optional if the popup is a `listbox`. If the combobox popup element is a `tree`, `grid`, or `dialog` (anything other than a `listbox`), the `aria-haspopup` attribute is required. The value of `aria-haspopup` must be either the `tree`, `grid`, `dialog`, or `listbox` role. Note that for this property, `true` means `menu`, so make sure that the value corresponds to the role of the popup, not a Boolean value. 

> When the popup is displayed, the [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls) attribute on the combobox element is set to the {{HTMLattrXRef('id')}} of the `listbox`, `tree`, `grid`, or `dialog` popup element of the combobox. This is how the relationship between the element with the `combobox` role and the popup it controls is indicated. (Note: In older ARIA specs, this was `aria-owns` rather than `aria-controls`, so you may see `aria-owns` in older combobox implementations.)

Reviewer: ```suggestion
When combobox's popup is displayed, ensure the [`aria-controls`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls) attribute on the combobox element is set to the {{HTMLattrXRef('id')}} of the popup `listbox`, `tree`, `grid`, or `dialog` element. This is how the relationship between the element with the `combobox` role and the popup it controls is indicated. (Note: In older ARIA specs, this was `aria-owns` rather than `aria-controls`, so you may see `aria-owns` in older combobox implementations. Such implementations should be updated to conform to the revised standard.)
```

> If the UI includes an visible control, such as an icon, that allows the visibility of the popup to be controlled via pointer and touch events, that control should be a {{HTMLElement('button')}}, {{HTMLElement('input')}} of type `button`, or a [`button`](/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) role element with a {{HTMLattrXRef('tabindex')}} of `-1`. The button should be focusable but not included in keyboard tab sequence. It should also not be a descendant of the element with role `combobox`. 

Reviewer: ```suggestion
If the combobox UI includes a visible control, such as an icon, that allows the visibility of the popup to be controlled via pointer and touch events, that control should be a {{HTMLElement('button')}}, {{HTMLElement('input')}} of type `button`, or a [`button`](/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) role element with a {{HTMLattrXRef('tabindex')}} of `-1`. Doing so will allow the button to be focusable but not included in keyboard tab sequence. It must not be a descendant of the element with role `combobox`. 
```

To be keyboard accessible, keyboard support for moving focus between the `combobox` input field element and elements contained in the popup `listbox`, `tree`, `grid`, or `dialog`, must be programmed in.  One common convention is that <kbd>Down Arrow</kbd> moves focus from the input to the first focusable descendant of the popup element. 

> The [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-activedescendant) property can be used to identify the currently active element when DOM focus is on a combobox. If the popup element supports `aria-activedescendant`, according to WAI-ARIA specification, it is OK to program updated values of the combobox element's `aria-activedescendant` property when a descendant of the popup element is active while focus remains on the combobox element instead of moving focus. 

Reviewer: ```suggestion
The [`aria-activedescendant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-activedescendant) property can be used to identify the currently active element of the combobox popup, for instance an `option` within a popup `listbox`, for implementations where DOM focus remains on the combobox. If DOM focus does not remain on the combobox when its popup is invoked, but rather DOM focus moves into the popup, such as a dialog, then `aria-activedescendant` may not be necessary.
```

### Keyboard interactions

- <kbd>Down Arrow</kbd>
  - : Moves focus to the next option, or to the first option if none was selected.

- <kbd>Alt</kbd> + <kbd>Down Arrow</kbd> (Optional)
  - : If the popup is available but not displayed, displays the popup without moving focus.

- <kbd>Up Arrow</kbd>
  - : Moves focus to the previous option. Moving focus to the first option if focus was originally on the last option.

- <kbd>Alt</kbd> + <kbd>Up Arrow</kbd> (Optional)
  - : If the popup has focus, returns focus to the combobox, otherwise it closes the popup.

- <kbd>Enter</kbd>
  - : If the combobox is editable and an autocomplete suggestion is selected in the popup, accepts the suggestion either by placing the input cursor at the end of the accepted value in the combobox or by performing a default action on the value. For example, in a messaging application, the default action may be to add the accepted value to a list of message recipients and then clear the combobox so the user can add another recipient.

## Examples

```html
<label for="jokes">Pick what type of jokes you like</label>
<div class="combo-wrap">

>   <input type="text" id="jokes" role="combobox" aria-controls="joketypes" aria-autocomplete="list" aria-expanded="false" data-active-option="item1" aria-activedescendant="item1">

Reviewer: since the combobox is in the collapsed state, activedescendant should be the empty string.
```suggestion
  <input type="text" id="jokes" role="combobox" aria-controls="joketypes" aria-autocomplete="list" aria-expanded="false" data-active-option="item1" aria-activedescendant="">
```

  <span aria-hidden="true" data-trigger="multiselect"></span>

>   <ul id="joketypes" role="listbox">

Reviewer: It is arguable how important naming a listbox is when it is invoked via a combobox with an accessible name, but since the ARIA specification presently does not differentiate between an in-page `listbox` vs a `listbox` popup for whether a name is required or not, a name should be added to mitigate against authors following this example and then getting conformance checker errors.
```suggestion
  <ul id="joketypes" role="listbox" aria-label="Jokes">
```
